File: app/src/main/java/com/iamkatrechko/clipboardmanager/fragment/ClipsListFragment.java
package com.iamkatrechko.clipboardmanager.fragment;

import android.app.Activity;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.iamkatrechko.clipboardmanager.DialogManager;
import com.iamkatrechko.clipboardmanager.R;
import com.iamkatrechko.clipboardmanager.activity.ClipEditActivity;
import com.iamkatrechko.clipboardmanager.activity.DeveloperActivity;
import com.iamkatrechko.clipboardmanager.activity.SearchActivity;
import com.iamkatrechko.clipboardmanager.adapter.ClipsCursorAdapter;
import com.iamkatrechko.clipboardmanager.util.UtilPreferences;

import static com.iamkatrechko.clipboardmanager.data.DatabaseDescription.Clip;

/**
 * Основной фрагмент экрана со списком заметок
 * @author iamkatrechko
 *         Date: 01.11.2016
 */
public class ClipsListFragment extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {

    /** Идентификатор загрузчика заметок по категории */
    private static final int CLIPS_BY_CATEGORY_LOADER = 1;

    /** Адаптер списка заметок */
    private ClipsCursorAdapter clipsCursorAdapter;
    /** Виджет списка заметок */
    private RecyclerView recyclerView;

    /** Текущая выбранная категория */
    private long currentCategoryId = 2;
    /** Включен ли режим выделения */
    private boolean isContextMenu;
    /** Количество выделенных элементов */
    private int selectedCount = 0;

    /**
     * Возвращает новый экземпляр фрагмента
     * @return новый экземпляр фрагмента
     */
    public static ClipsListFragment newInstance() {
        return new ClipsListFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_main, container, false);

        recyclerView = (RecyclerView) v.findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity().getBaseContext()));
        recyclerView.setHasFixedSize(true);

        clipsCursorAdapter = new ClipsCursorAdapter(getActivity(), new ClipsCursorAdapter.ClipClickListener() {

            @Override
            public void onClick(long clipId) {
                Intent i = ClipEditActivity.newIntent(getActivity(),
                        Clip.buildClipUri(clipId));
                startActivity(i);
            }

            @Override
            public void onSelectedChange(boolean isSelectedMode, int selectedCount) {
                isContextMenu = isSelectedMode;
                getActivity().invalidateOptionsMenu();
                ClipsListFragment.this.selectedCount = selectedCount;
            }
        });
        clipsCursorAdapter.setEmptyView(v.findViewById(R.id.linearEmpty));
        recyclerView.setAdapter(clipsCursorAdapter);

        showClipsByCategoryId(2);
        return v;
    }

    /**
     * Отображает список заметок по категории
     * @param categoryId идентификатор категории
     */
    public void showClipsByCategoryId(long categoryId) {
        clipsCursorAdapter.resetSelectMode();
        Bundle bundle = new Bundle();
        bundle.putLong("categoryId", categoryId);
        currentCategoryId = categoryId;
        getLoaderManager().restartLoader(CLIPS_BY_CATEGORY_LOADER, bundle, this);
    }

    /** Открывает экран создания новой заметки */
    public void addNewClip() {
        Intent i = ClipEditActivity.newIntent(getActivity(), null);
        startActivity(i);
    }

    /**
     * Меняет значок ToolBar"а (звездочку) в зависимости от того,
     * отображаются ли только избранные записи, или нет
     * @param itemStar           Элемент ToolBar'а
     * @param isOnlyFavoriteShow Отображаются только избранные записи
     */
    private void changeToolbarItemIcon(MenuItem itemStar, boolean isOnlyFavoriteShow) {
        if (isContextMenu) return;
        if (isOnlyFavoriteShow) {
            itemStar.setIcon(R.drawable.ic_star_white);
        } else {
            itemStar.setIcon(R.drawable.ic_star_border_white);
        }
    }

    /** Нажатие кнопки "назад" */
    public void onBackPressed() {
        if (isContextMenu) {
            clipsCursorAdapter.resetSelectMode();
        } else {
            getActivity().finish();
        }
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        Log.d("Fragment", "onCreateLoader");

        switch (id) {
            /*case CLIPS_LOADER:
                Log.d("Fragment", "onCreateLoader2");
                return new CursorLoader(getActivity(),
                        Clip.CONTENT_URI, // Uri таблицы contacts
                        null, // все столбцы
                        null, // все записи
                        null, // без аргументов
                        Clip._ID + " DESC"); // сортировка*/
            case CLIPS_BY_CATEGORY_LOADER:
                String orderType = UtilPreferences.getOrderType(getActivity());
                String orderQuery = null;
                if (orderType.equals("3")) {
                    orderQuery = Clip.COLUMN_DATE;
                }
                if (orderType.equals("2")) {
                    orderQuery = Clip.COLUMN_DATE + " DESC";
                }
                if (orderType.equals("1")) {
                    orderQuery = Clip._ID;
                }
                long categoryId = args.getLong("categoryId");
                boolean isOnlyFavoriteShow = UtilPreferences.isShowOnlyFavorite(getActivity());
                String onlyFavorite = isOnlyFavoriteShow ? " and " + Clip.COLUMN_IS_FAVORITE + " = 1" : "";
                return new CursorLoader(getActivity(),
                        Clip.CONTENT_URI, // Uri таблицы contacts
                        null, // все столбцы
                        Clip.COLUMN_CATEGORY_ID + "=?" + onlyFavorite, // все записи
                        new String[]{String.valueOf(categoryId)}, // без аргументов
                        orderQuery); // сортировка
            default:
                return null;
        }
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        clipsCursorAdapter.setCursor(data);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader) {
        clipsCursorAdapter.setCursor(null);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(!isContextMenu ? R.menu.menu_main : R.menu.menu_main_context, menu);
        if (selectedCount == 0) {
            getActivity().setTitle(R.string.app_name);
        } else {
            getActivity().setTitle("" + selectedCount);
            menu.findItem(R.id.action_split).setVisible(selectedCount > 1);
        }

        boolean showOnlyFavorite = UtilPreferences.isShowOnlyFavorite(getActivity());
        MenuItem itemStar = menu.findItem(R.id.action_show_favorites);
        changeToolbarItemIcon(itemStar, showOnlyFavorite);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            // Окно поиска
            case R.id.action_search:
                getActivity().startActivity(new Intent(getActivity(), SearchActivity.class));
                break;
            // Настройка сортировки
            case R.id.action_set_order:
                DialogManager.showDialogSetOrderType(this);
                break;
            // Показывать только избранные
            case R.id.action_show_favorites:
                boolean isOnly = UtilPreferences.isShowOnlyFavorite(getActivity());
                changeToolbarItemIcon(item, !isOnly);
                UtilPreferences.setShowOnlyFavorite(getActivity(), !isOnly);
                showClipsByCategoryId(currentCategoryId);
                break;
            // Меню разработчика
            case R.id.action_developer_menu:
                Intent i = new Intent(getActivity(), DeveloperActivity.class);
                getActivity().startActivity(i);
                break;
            // Удалить выделенные записи
            case R.id.action_delete:
                DialogManager.showDialogDeleteConfirm(this);
                break;
            // Соединить выделенные записи
            case R.id.action_split:
                DialogManager.showDialogSplitClips(this);
                break;
            // Поделиться выделенными записями
            case R.id.action_share:
                clipsCursorAdapter.shareItems();
                break;
            // Сменить категорию выделенных записей
            case R.id.action_change_category:
                DialogManager.showDialogChangeCategory(this);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK && requestCode == DialogManager.DIALOG_SPLIT_CLIPS) {
            String splitChar = data.getStringExtra("splitChar");
            boolean deleteOldClips = data.getBooleanExtra("deleteOldClips", false);

            clipsCursorAdapter.splitItems(splitChar, deleteOldClips);
        }
        if (resultCode == Activity.RESULT_OK && requestCode == DialogManager.DIALOG_CHANGE_CATEGORY) {
            long categoryId = data.getLongExtra("categoryId", 0);

            clipsCursorAdapter.changeCategory(categoryId);
        }
        if (resultCode == Activity.RESULT_OK && requestCode == DialogManager.DIALOG_DELETE_CONFIRM) {
            boolean delete = data.getBooleanExtra("delete", false);
            if (delete)
                clipsCursorAdapter.deleteSelectedItems();
        }
        if (resultCode == Activity.RESULT_OK && requestCode == DialogManager.DIALOG_SET_ORDER_TYPE) {
            String orderType = data.getStringExtra("orderType");
            UtilPreferences.setOrderType(getActivity(), orderType);
            showClipsByCategoryId(currentCategoryId);
        }
    }
}

File: app/src/main/java/com/iamkatrechko/clipboardmanager/view/fragment/ClipsListFragment.kt
package com.iamkatrechko.clipboardmanager.view.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.*
import android.widget.Toast
import com.iamkatrechko.clipboardmanager.App
import com.iamkatrechko.clipboardmanager.R
import com.iamkatrechko.clipboardmanager.data.repository.ClipboardRepository
import com.iamkatrechko.clipboardmanager.databinding.FragmentMainBinding
import com.iamkatrechko.clipboardmanager.domain.ClipsHelper
import com.iamkatrechko.clipboardmanager.domain.loader.callback.ClipsLoaderCallback
import com.iamkatrechko.clipboardmanager.domain.param.ClipParam
import com.iamkatrechko.clipboardmanager.domain.param.values.OrderType
import com.iamkatrechko.clipboardmanager.domain.util.IntentUtils
import com.iamkatrechko.clipboardmanager.domain.util.PrefsManager
import com.iamkatrechko.clipboardmanager.view.DialogManager
import com.iamkatrechko.clipboardmanager.view.activity.ClipEditActivity
import com.iamkatrechko.clipboardmanager.view.activity.ClipViewActivity
import com.iamkatrechko.clipboardmanager.view.activity.DeveloperActivity
import com.iamkatrechko.clipboardmanager.view.activity.SearchActivity
import com.iamkatrechko.clipboardmanager.view.adapter.ClipsAdapter
import com.iamkatrechko.clipboardmanager.view.base.BaseFragment
import com.iamkatrechko.clipboardmanager.view.dialog.DialogChangeCategory
import com.iamkatrechko.clipboardmanager.view.dialog.DialogDeleteConfirm
import com.iamkatrechko.clipboardmanager.view.dialog.DialogSplitClips
import com.iamkatrechko.clipboardmanager.view.extension.showToast
import com.iamkatrechko.clipboardmanager.view.presenter.ClipsListPresenter
import com.iamkatrechko.clipboardmanager.view.presenter.ClipsListView

/**
 * Основной фрагмент экрана со списком заметок
 * @author iamkatrechko
 *         Date: 01.11.2016
 */
class ClipsListFragment : BaseFragment<FragmentMainBinding>(), ClipsListView {

    override val layoutId: Int
        get() = R.layout.fragment_main

    /** Презентер экрана */
    private lateinit var presenter: ClipsListPresenter
    /** Виджет списка заметок  */
    private lateinit var recyclerView: RecyclerView

    /** Менеджер буфера обмена */
    private val clipManager = App.clipManager
    /** Текущая выбранная категория  */
    private var currentCategoryId: Long? = -1
    /** Включен ли режим выделения  */
    private var isContextMenu: Boolean = false
    /** Количество выделенных элементов  */
    private var selectedCount = 0
    /** Репозиторий для работы с базой записей  */
    private var repository = ClipboardRepository.getInstance()
    /** Слушатель для адаптера списка */
    private var listener = object : ClipsAdapter.ClipClickListener {

        override fun onClick(clipId: Long) {
            startActivity(ClipEditActivity.newIntent(requireContext(), clipId))
        }

        override fun onSelectedChange(isSelectedMode: Boolean, selectedCount: Int) {
            isContextMenu = isSelectedMode
            this@ClipsListFragment.selectedCount = selectedCount
            activity?.invalidateOptionsMenu()
        }
    }
    /** Адаптер списка заметок  */
    private var clipsAdapter: ClipsAdapter = ClipsAdapter(listener)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        currentCategoryId = arguments?.getLong(KEY_ARGUMENT_CATEGORY_ID, -1L)
        currentCategoryId = if (currentCategoryId == -1L) null else currentCategoryId
        presenter = ClipsListPresenter()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        recyclerView = binding.recyclerView
        clipsAdapter.setEmptyView(binding.linearEmpty)

        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = clipsAdapter

        clipsAdapter.onMoreClickListener = ::showPopupMenu
        showClipsByCategoryId(currentCategoryId)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter.onAttach(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.onDetach()
    }

    override fun closeScreen() {
        activity?.finish()
    }

    /**
     * Отображает список заметок по категории
     * @param categoryId идентификатор категории
     */
    fun showClipsByCategoryId(categoryId: Long?) {
        currentCategoryId = categoryId
        val order = PrefsManager.getInstance().clipsOrderType
        val onlyFav = PrefsManager.getInstance().isShowOnlyFavorite
        clipsAdapter.resetSelectMode()
        val bundle = Bundle().apply {
            putParcelable(ClipsLoaderCallback.KEY_LOADER_PARAMS, ClipParam(categoryId = categoryId, order = order, onlyFav = onlyFav))
        }
        loaderManager.restartLoader(ClipsLoaderCallback.MAIN_CLIPS_LOADER, bundle,
                ClipsLoaderCallback(requireContext(), clipsAdapter::setClips))
    }

    override fun openClipViewScreen(clipId: Int) {
        startActivity(ClipViewActivity.newIntent(requireContext(), clipId.toLong()))
    }

    override fun openClipEditScreen(clipId: Int) {
        startActivity(ClipEditActivity.newIntent(requireContext(), clipId.toLong()))
    }

    private fun showPopupMenu(view: View, pos: Int, clipId: Long) {
        val popupMenu = PopupMenu(requireContext(), view)
        popupMenu.inflate(R.menu.menu_popup_clip_options)
        popupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_copy -> {
                    clipManager.toClipboard(repository.getClip(clipId)?.text.orEmpty())
                    clipsAdapter.notifyDataSetChanged()
                }
                R.id.action_copy_close -> presenter.onCopyAndCloseClicked(clipId.toInt())
                R.id.action_view -> presenter.onClipViewClicked(clipId.toInt())
                R.id.action_edit -> presenter.onClipEditClicked(clipId.toInt())
                R.id.action_delete -> {
                    repository.deleteClip(clipId)
                    clipsAdapter.notifyItemRemoved(pos)
                }
                R.id.action_share -> {
                    IntentUtils.sendMail(requireContext(), repository.getClip(clipId)?.text.orEmpty())
                }
            }
            return@setOnMenuItemClickListener true
        }
        popupMenu.show()
    }

    /**
     * Меняет значок ToolBar"а (звездочку) в зависимости от того,
     * отображаются ли только избранные записи, или нет
     * @param itemStar           Элемент ToolBar'а
     * @param isOnlyFavoriteShow Отображаются только избранные записи
     */
    private fun changeToolbarItemIcon(itemStar: MenuItem?, isOnlyFavoriteShow: Boolean) {
        if (isContextMenu) return
        itemStar?.setIcon(if (isOnlyFavoriteShow) R.drawable.ic_star_white else R.drawable.ic_star_border_white)
    }

    /** Нажатие кнопки "назад"  */
    fun onBackPressed() {
        if (isContextMenu) {
            clipsAdapter.resetSelectMode()
        } else {
            activity!!.finish()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(if (!isContextMenu) R.menu.menu_main else R.menu.menu_main_context, menu)
        if (selectedCount == 0) {
            activity?.setTitle(R.string.app_name)
        } else {
            activity?.title = getString(R.string.selected_count, selectedCount.toString())
            menu.findItem(R.id.action_split)?.isVisible = selectedCount > 1
        }

        val showOnlyFavorite = PrefsManager.getInstance().isShowOnlyFavorite
        val itemStar = menu.findItem(R.id.action_show_favorites)
        changeToolbarItemIcon(itemStar, showOnlyFavorite)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            // Окно поиска
            R.id.action_search -> startActivity(Intent(activity, SearchActivity::class.java))
            // Настройка сортировки
            R.id.action_set_order -> DialogManager.showDialogSetOrderType(this)
            // Показывать только избранные
            R.id.action_show_favorites -> {
                val isOnly = PrefsManager.getInstance().isShowOnlyFavorite
                changeToolbarItemIcon(item, !isOnly)
                PrefsManager.getInstance().isShowOnlyFavorite = !isOnly
                showClipsByCategoryId(currentCategoryId)
            }
            // Меню разработчика
            R.id.action_developer_menu -> {
                startActivity(Intent(requireContext(), DeveloperActivity::class.java))
            }
            // Удалить выделенные записи
            R.id.action_delete -> DialogManager.showDialogDeleteConfirm(this)
            // Соединить выделенные записи
            R.id.action_split -> DialogManager.showDialogSplitClips(this)
            // Поделиться выделенными записями
            R.id.action_share -> {
                val shareText = ClipsHelper.joinToString(
                        clipsAdapter.getSelectedIds(),
                        PrefsManager.getInstance().clipSplitChar)
                IntentUtils.sendMail(requireContext(), shareText)
            }
            // Сменить категорию выделенных записей
            R.id.action_change_category -> DialogManager.showDialogChangeCategory(this)
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != Activity.RESULT_OK || data == null) {
            return
        }
        if (requestCode == DialogManager.DIALOG_SPLIT_CLIPS) {
            val splitChar = data.getStringExtra(DialogSplitClips.KEY_SPLIT_CHAR)
            if (clipsAdapter.getSelectedIds().isNotEmpty()) {
                val deleteOldClips = data.getBooleanExtra(DialogSplitClips.KEY_IS_DELETE_OLD_CLIPS, false)
                ClipsHelper.joinAndDelete(clipsAdapter.getSelectedIds(), splitChar, deleteOldClips)
                clipsAdapter.resetSelectMode()
                showToast(R.string.splited)
            } else {
                showToast(R.string.select_cancel)
            }
        }
        if (requestCode == DialogManager.DIALOG_CHANGE_CATEGORY) {
            val categoryId = data.getLongExtra(DialogChangeCategory.KEY_CATEGORY_ID, 0)
            ClipsHelper.changeCategory(clipsAdapter.getSelectedIds(), categoryId)
        }
        if (requestCode == DialogManager.DIALOG_DELETE_CONFIRM) {
            val delete = data.getBooleanExtra(DialogDeleteConfirm.KEY_IS_DELETE, false)
            if (delete) {
                repository.deleteClips(clipsAdapter.getSelectedIds())
                clipsAdapter.resetSelectMode()
            }
        }
        if (requestCode == DialogManager.DIALOG_SET_ORDER_TYPE) {
            val pos = data.getIntExtra("orderType", 0)
            PrefsManager.getInstance().clipsOrderType = OrderType.values()[pos]
            showClipsByCategoryId(currentCategoryId)
        }
    }

    companion object {

        /** Ключ аргумента. Идентификатор текущей категории */
        private const val KEY_ARGUMENT_CATEGORY_ID = "CATEGORY_ID"

        /**
         * Возвращает новый экземпляр фрагмента
         * @param [categoryId] идентификатор категории
         * @return новый экземпляр фрагмента
         */
        fun newInstance(categoryId: Long? = null): ClipsListFragment {
            val fragment = ClipsListFragment()

            fragment.arguments = Bundle().apply {
                putLong(KEY_ARGUMENT_CATEGORY_ID, categoryId ?: -1)
            }

            return fragment
        }
    }
}
